### PR TITLE
Fix EOF (^D) handling in devpts (partially addr #9333)

### PR DESCRIPTION
**Pull request by thundergolfer**
This fixes the ctrl-d part of #9333.

Disclaimer: new to this codebase.

### CODE REVIEW

File: pkg/sentry/fsimpl/devpts/line_discipline.go [hunk 1]
```diff
@@ -212,9 +212,15 @@ func (l *lineDiscipline) inputQueueRead(ctx context.Context, dst usermem.IOSeque
 			l.replicaWaiter.Notify(waiter.ReadableEvents)
 		}
 		return n, nil
-	} else if notifyEcho {
-		l.masterWaiter.Notify(waiter.ReadableEvents)
+	} else {
```
Comment (nlacasse) [inline pkg/sentry/fsimpl/devpts/line_discipline.go:215]: Nit: `else` is not needed in this case, since the preceeding `if` block returns unconditionally.

**Review by nlacasse — CHANGES_REQUESTED** [review]
Thanks for the fix, and the nice test. One minor nit.

**Comment by ayushr2** [comment]
Could you squash the two commits? Otherwise these will appear as two separate commits on the master branch after merge.

**Comment by thundergolfer** [comment]
Oh sure, I thought I could squash just before merge. But I'll squash them now and force push.

**Comment by ayushr2** [comment]
We don't squash and merge from Github UI (unfortunately...). It needs be go through our internal system (via the "ready to pull" tag). [Copybara](https://github.com/google/copybara) will clone your PR as a new PR and merge that and close this. And in the process, it just pushes the commits from this branch onto master. So we request the commits are squashed (unless they are meant to be logically separated).

**Comment by thundergolfer** [comment]
Make sense. Thanks for the explanation 👌

**Review by nlacasse — APPROVED** [review]

**Comment by nlacasse** [comment]
This seems to break the `PtyTest.TermiosONLCR` test:
```
[ RUN      ] PtyTest.TermiosONLCR
test/syscalls/linux/pty.cc:466: Failure
Value of: ReadFd(fd.get(), &c, 1)
Expected: -1 (failure), with errno PosixError(errno=11 Resource temporarily unavailable)
  Actual: 0 (of type long)
[  FAILED  ] PtyTest.TermiosONLCR (5 ms)
```

Full logs [here](https://storage.googleapis.com/gvisor-buildkite/018a666d-8936-49f2-9f4e-076ab5c2df43/018a666d-d375-4eb9-a07f-be5ab3fa5e63/home/agent-090523-165034/.cache/bazel/_bazel_agent-090523-165034/99920da594400bd42ae3b33af614153e/execroot/__main__/bazel-out/k8-fastbuild/testlogs/test/syscalls/pty_test_runsc_ptrace_shared/shard_1_of_4/test.log)

**Comment by thundergolfer** [comment]
Oh of course, this behavior is only valid in canonical mode, and that test disables that flag. Pushed a fix.

File: pkg/sentry/fsimpl/devpts/line_discipline.go [hunk 2]
```diff
@@ -212,9 +212,14 @@ func (l *lineDiscipline) inputQueueRead(ctx context.Context, dst usermem.IOSeque
 			l.replicaWaiter.Notify(waiter.ReadableEvents)
 		}
 		return n, nil
-	} else if notifyEcho {
+	}
+	if notifyEcho {
 		l.masterWaiter.Notify(waiter.ReadableEvents)
 	}
+	if !pushed && l.termios.LEnabled(linux.ICANON) {
```
Comment (nlacasse) [inline pkg/sentry/fsimpl/devpts/line_discipline.go:219]: We have to hold `l.termiosMu` to access `l.termios`.

I would set `isCanon := t.ermios.LEnabled(linux.ICANON)`  while the lock is held on line 200, and then reference `isCanon` in this conditional.

**Review by nlacasse — CHANGES_REQUESTED** [review]

**Review by nlacasse — APPROVED** [review]

**Comment by nlacasse** [comment]
This is merged!  Thanks for the PR.

**Pull request merged**